src/lib/homestore_backend/hs_cp_callbacks.cpp

```diff
@@ -0,0 +1,59 @@
+/*********************************************************************************
+ * Modifications Copyright 2017-2019 eBay Inc.
+ *
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ *********************************************************************************/
+#include <vector>
+#include <homestore/homestore.hpp>
+#include "hs_homeobject.hpp"
+
+using homestore::CP;
+using homestore::CPCallbacks;
+using homestore::CPContext;
+
+namespace homeobject {
+
+std::unique_ptr< CPContext > HSHomeObject::MyCPCallbacks::on_switchover_cp(CP* cur_cp, CP* new_cp) { return nullptr; }
+
+// when cp_flush is called, it means that all the dirty candidates are already in the dirty list.
+// new dirty candidates will arrive on next cp's context.
+folly::Future< bool > HSHomeObject::MyCPCallbacks::cp_flush(CP* cp) {
+    std::vector< HSHomeObject::HS_PG* > dirty_pg_list;
+    dirty_pg_list.reserve(home_obj_._pg_map.size());
+    {
+        std::shared_lock lock_guard(home_obj_._pg_lock);
```

Comment:
Is there a reason you choose to hold this _pg_lock? The original idea is cp_flush won't need to take this lock (because BlobManager is going to take the lock for every blob put) so that no to block I/O during cp_flush. Or do you intend not to take that lock during blob put?

Reply:
We are not taking lock while during /O during cp_flush, Under the lock, it only takes the durables, update the in-mem version of the pg_sb and then unlocks and persist them. In the critical data path, we could avoid putting the pg into an unordered_map which we need to put under the lock. At present, we just update each PG's atomics and then gather all PGs metadata during cp_flush.

Reply:
c++20 has initialize for ranged-for loops
```c++
for (auto lock_guard = shared_lock(home_obj_._pg_lock); auto const& [id, pg] ...) {
   ...
}
```

Reply:
That lock can be removed actually. Personally I think we should keep cp out of the hosting consumers as much as possible. People who works on PG (HS_PG) doesn't need to look into what is meaning of hs_pg->pg_sb_, to understand it is actually only touched by cp routine. So why not keep anything need to be flushed in a dedicated class and let that class do the dirty buff flush? 

With your changes here, we can move hs_pg->pg_sb_ into HomeObjectCPContext, right?

Reply:
this is a minor comment above, I am okay to merge as it is. 
If you still are considering doing this, a requirement from HomeStore (as we discussed yesterday) is we need to allow this pg_sb_ to be persistent outside of cp_flush as well.